### DISCUSSION TITLE
Does TorchIO have its way of controlling randomness/reproducibility?

### DISCUSSION THREAD

**Discussion by dzenanz** [TorchIO-project/torchio#766] [Q&A] [answered]
In order to have the same results each time my code is run, is [general pytorch discussion](https://pytorch.org/docs/stable/notes/randomness.html) the most valid, or does TorchIO has its own (easier) way of controlling it?

**Comment by romainVala** [comment]
Hi ,
I think all random state of torchio Random transform comes from  numpy , so using 
np.random.seed(0) at the begining of your code, should be ok.

but the dataselection (and randomness) comes from  torch dataloader, so you should also add
torch.manual_seed(0)

**Reply by dzenanz** [reply]
Is `worker_init_fn=seed_worker` needed when `num_workers` is greater than 0?

**Comment by fepegar** [comment]
@romainVala 

> Hi ,
I think all random state of torchio Random transform comes from numpy , so using
np.random.seed(0) at the begining of your code, should be ok.
but the dataselection (and randomness) comes from torch dataloader, so you should also add
torch.manual_seed(0)

NumPy is not used for random operations, so `torch.manual_seed` is what should be used.

---

@dzenanz 

> Is worker_init_fn=seed_worker needed when num_workers is greater than 0?

I don't think so, but I'm not sure.

> Is worker_init_fn=seed_worker needed when num_workers is greater than 0?

No, I don't think so.

**Reply by fepegar** [reply]
FYI:

```python
import torch


class Dataset:
    def __len__(self):
        return 4

    def __getitem__(self, _):
        return torch.rand(1)


torch.manual_seed(0)

loader = torch.utils.data.DataLoader(
    Dataset(),
    num_workers=2,
    batch_size=4,
)

print(next(iter(loader)))
```

Running the script above:

```shell
$ python script.py     
tensor([[0.7821],
        [0.0536],
        [0.9888],
        [0.1949]])
$ python script.py                  
tensor([[0.7821],
        [0.0536],
        [0.9888],
        [0.1949]])
```

**Reply by romainVala** [reply]
> NumPy is not used for random operations, so `torch.manual_seed` is what should be used.

[here](https://github.com/fepegar/torchio/blob/4e4be011761140d57d13be38d6182102f75d216b/torchio/data/queue.py#L238) is a call to random which is neither torch nor numpy, but not sure how to control this seed ... ?

**Reply by dzenanz** [reply]
`random.seed(10)` should do it?

Yesterday, I added
```py
np.random.seed(30101983)
torch.manual_seed(30101983)
```
I ran my training overnight, now I am re-running it to see whether I get the same results again.

**Reply by dzenanz** [reply]
With `testTorchRepro.py`:
```py
import torch

class Dataset:
    def __len__(self):
        return 16

    def __getitem__(self, _):
        return torch.rand(1)

if __name__ == "__main__":
    torch.manual_seed(30101983)
    loader = torch.utils.data.DataLoader(
        Dataset(),
        num_workers=4,
        batch_size=16,
    )
    print(next(iter(loader)))
```
generated numbers are reproducible from run to run.

**Reply by fepegar** [reply]
> here is a call to random which is neither torch nor numpy, but not sure how to control this seed ... ?

Oh, good catch. That'd be `random.seed`, as @dzenanz said. But for consistency, I think we should use `torch` there. I'll open an issue.

**Reply by fepegar** [reply]
Hmm for some reason, the code below generates consistent results. I would expect it to change as it uses `random`, but it doesn't. I guess the issue won't be needed...

```python
import torch
import torchio as tio

torch.manual_seed(0)

dataset = tio.datasets.OrganMNIST3D(split='test')
queue = tio.Queue(
    dataset,
    100,
    5,
    tio.UniformSampler(12),
)
loader = torch.utils.data.DataLoader(
    queue,
    num_workers=2,
    batch_size=4,
    shuffle=True,
)

print(next(iter(loader)))
```

**Reply by dzenanz** [reply]
> Yesterday, I added
> ```python
> np.random.seed(30101983)
> torch.manual_seed(30101983)
> ```
> 
> I ran my training overnight, now I am re-running it to see whether I get the same results again.

Numbers are similar now, but not exactly the same. I will try to also add `random.seed(30101983)`.

**Reply by fepegar** [reply]
Thanks for reporting. Did you also used `torch.use_deterministic_algorithms`?

**Reply by dzenanz** [reply]
No. I will try that.

Adding `random.seed()` improves reproducibility, the numbers are noticeably closer but still not exactly the same.

**Reply by dzenanz** [reply]
That did it. Adding the following code to my program made it fully reproducible across runs:
```py
random.seed(30101983)
np.random.seed(30101983)
torch.manual_seed(30101983)
torch.use_deterministic_algorithms(True)
```

**Reply by dzenanz** [reply]
Thanks both @fepegar and @romainVala!

**Comment by dzenanz** [comment, marked as answer]
TL;DR version: Adding this made the program fully reproducible across runs:
```py
random.seed(30101983)
np.random.seed(30101983)
torch.manual_seed(30101983)
torch.use_deterministic_algorithms(True)
```

**Reply by fepegar** [reply]
Thanks, @dzenanz! I have opened a PR to remove the shuffle using `random`, so, in theory, only the last two lines should be relevant for TorchIO:
- #776
